airflow/utils/log/log_reader.py

```diff
@@ -0,0 +1,119 @@
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import logging
+from typing import Any, Dict, Iterator, List, Optional, Tuple
+
+from cached_property import cached_property
+
+from airflow.configuration import conf
+from airflow.models import TaskInstance
+from airflow.utils.helpers import render_log_filename
+
+
+class TaskLogReader:
+    """ Task log reader"""
+
+    def read_log_chunks(self, ti: TaskInstance, try_number: Optional[int],
+                        metadata) -> Tuple[List[str], Dict[str, Any]]:
+        """
+        Reads chunks of Task Instance logs.
+
+        :param ti: The taskInstance
+        :type ti: TaskInstance
+        :param try_number: If provided, logs for the given try will be returned.
+            Otherwise, logs from all attempts are returned.
+        :type try_number: Optional[int]
+        :param metadata: A dictionary containing information about how to read the task log
+        :type metadata: dict
+        :rtype: Tuple[List[str], Dict[str, Any]]
+
+        The following is an example of how to use this method to read log:
+
+        .. code-block:: python
+
+            logs, metadata = task_log_reader.read_log_chunks(ti, try_number, metadata)
+            logs = logs[0] if try_number is not None else logs
+
+        where task_log_reader is an instance of TaskLogReader. The metadata will always
+        contain information about the task log which can enable you read logs to the
+        end.
+        """
+
+        logs, metadatas = self.log_handler.read(ti, try_number, metadata=metadata)
+        metadata = metadatas[0]
+        return logs, metadata
+
+    def read_log_stream(self, ti: TaskInstance, try_number: Optional[int],
+                        metadata: dict) -> Iterator[str]:
+        """
+        Used to continuously read log to the end
+
+        :param ti: The Task Instance
+        :type ti: TaskInstance
+        :param try_number: the task try number
+        :type try_number: Optional[int]
+        :param metadata: A dictionary containing information about how to read the task log
+        :type metadata: dict
+        :rtype: Iterator[str]
+        """
+
+        if try_number is None:
+            next_try = ti.next_try_number
+            try_numbers = list(range(1, next_try))
+        else:
+            try_numbers = [try_number]
+        for current_try_number in try_numbers:
+            metadata.pop('end_of_log', None)
+            metadata.pop('max_offset', None)
+            metadata.pop('offset', None)
+            while 'end_of_log' not in metadata or not metadata['end_of_log']:
+                logs, metadata = self.read_log_chunks(ti, current_try_number, metadata)
+                yield "\n".join(logs) + "\n"
```

Comment:
What would you say to use variable?
```
end_of_log = metadata.get("end_of_log")
```

Reply:
The key may not exist and if it exist, is a boolean